File: Example/Tests/ContainerTests.m
//
//  ContainerTests.m
//  Clappr
//
//  Created by Gustavo Barbosa on 12/11/14.
//  Copyright (c) 2014 globo.com. All rights reserved.
//

#import <Clappr/Clappr.h>

SPEC_BEGIN(Container)

describe(@"Container", ^{

    __block CLPContainer *container;
    __block CLPPlayback *playback;

    context(@"Instantiation", ^{

        it(@"cannot be instatiated without a playback", ^{
            [[theBlock(^{
                [CLPContainer new];
            }) should] raiseWithName:NSInternalInconsistencyException];
        });

        it(@"should receive a playback in the constructor", ^{
            playback = [CLPPlayback new];
            container = [[CLPContainer alloc] initWithPlayback:playback];
            [[container.playback should] equal:playback];
        });
    });

    context(@"General", ^{

        beforeEach(^{
            playback = [CLPPlayback new];
            container = [[CLPContainer alloc] initWithPlayback:playback];
        });

        it(@"can be destroyed properly", ^{
            UIView *wrapperView = [UIView new];
            [wrapperView addSubview:container.view];

            __block NSString *containerName;
            [container once:CLPContainerEventDestroyed callback:^(NSDictionary *userInfo) {
                containerName = userInfo[@"name"];
            }];

            [[playback should] receive:@selector(destroy)];
            [container destroy];
            [[container.view.superview should] beNil];
            [[containerName should] equal:@"Container"];
        });

        it(@"should call playback's play method after its play method has been called", ^{
            [[playback should] receive:@selector(play)];
            [container play];
        });

        it(@"should call playback's pause method after its pause method has been called", ^{
            [[playback should] receive:@selector(pause)];
            [container pause];
        });
    });

    context(@"event binding", ^{

        beforeEach(^{
            playback = [CLPPlayback new];
            container = [[CLPContainer alloc] initWithPlayback:playback];
        });

        it(@"should trigger its event after listen to playback's progress event with the respective params", ^{
            // Given
            float expectedStartPosition = 0.7f, expectedEndPosition = 15.4f;
            NSTimeInterval expectedDuration = 10.0f;

            __block float startPosition, endPosition;
            __block NSTimeInterval duration;

            [container once:CLPContainerEventProgress callback:^(NSDictionary *userInfo) {
                startPosition = [userInfo[@"start_position"] floatValue];
                endPosition = [userInfo[@"end_position"] floatValue];
                duration = [userInfo[@"duration"] doubleValue];
            }];

            // When
            NSDictionary *userInfo = @{
                @"start_position": @(expectedStartPosition),
                @"end_position": @(expectedEndPosition),
                @"duration": @(expectedDuration)
            };
            [playback trigger:CLPPlaybackEventProgress userInfo:userInfo];

            // Then
            [[theValue(startPosition) should] equal:theValue(expectedStartPosition)];
            [[theValue(endPosition) should] equal:theValue(expectedEndPosition)];
            [[theValue(duration) should] equal:theValue(expectedDuration)];
        });

        it(@"should trigger its event after listen to playback's time updated event with the respective params", ^{
            // Given
            float expectedPosition = 10.3f;
            NSTimeInterval expectedDuration = 12.78;

            __block float position = 0.0;
            __block NSTimeInterval duration = 0.0;
            [container once:CLPContainerEventTimeUpdate callback:^(NSDictionary *userInfo) {
                position = [userInfo[@"position"] floatValue];
                duration = [userInfo[@"duration"] doubleValue];
            }];

            // When
            NSDictionary *userInfo = @{
                @"position": @(expectedPosition),
                @"duration":@(expectedDuration)
            };
            [playback trigger:CLPPlaybackEventTimeUpdated userInfo:userInfo];

            // Then
            [[theValue(position) should] equal:theValue(expectedPosition)];
            [[theValue(duration) should] equal:theValue(expectedDuration)];
        });

        it(@"should be ready after listen to playback's ready event", ^{
            [[theValue([container isReady]) should] beFalse];

            [playback trigger:CLPPlaybackEventReady];

            [[theValue([container isReady]) should] beTrue];
        });

        it(@"should trigger its event after listen to playback's buffering event", ^{
            __block BOOL eventWasTriggered = NO;
            [container once:CLPContainerEventBuffering callback:^(NSDictionary *userInfo) {
                eventWasTriggered = YES;
            }];

            [playback trigger:CLPPlaybackEventBuffering];

            [[theValue(eventWasTriggered) should] beTrue];
        });

        it(@"should trigger its event after listen to playback's buffer full event", ^{
            __block BOOL eventWasTriggered = NO;
            [container once:CLPContainerEventBufferFull callback:^(NSDictionary *userInfo) {
                eventWasTriggered = YES;
            }];

            [playback trigger:CLPPlaybackEventBufferFull];

            [[theValue(eventWasTriggered) should] beTrue];
        });

        it(@"should update its settings after listen to playback's settings update event", ^{
            [playback stub:@selector(settings) andReturn:@{@"foo": @"bar"}];

            [playback trigger:CLPPlaybackEventSettingsUdpdated];

            [[container.settings should] equal:@{@"foo": @"bar"}];
        });

        it(@"should trigger its event after listen to playback settings update event", ^{
            __block BOOL eventWasTriggered = NO;
            [container once:CLPContainerEventSettingsUpdated callback:^(NSDictionary *userInfo) {
                eventWasTriggered = YES;
            }];

            [playback trigger:CLPPlaybackEventSettingsUdpdated];

            [[theValue(eventWasTriggered) should] beTrue];
        });

        it(@"should trigger its event after listen to playback's loaded metadata event with the respective params", ^{
            NSTimeInterval expectedDuration = 20.0;

            __block NSTimeInterval duration = 0.0;
            [container once:CLPContainerEventLoadedMetadata callback:^(NSDictionary *userInfo) {
                duration = [userInfo[@"duration"] doubleValue];
            }];

            NSDictionary *userInfo = @{@"duration":@(expectedDuration)};
            [playback trigger:CLPPlaybackEventLoadedMetadata userInfo:userInfo];

            [[theValue(duration) should] equal:theValue(expectedDuration)];
        });

        it(@"should trigger its event after listen to playback's HD update event", ^{
            __block BOOL eventWasTriggered = NO;
            [container once:CLPContainerEventHighDefinitionUpdated callback:^(NSDictionary *userInfo) {
                eventWasTriggered = YES;
            }];

            [playback trigger:CLPPlaybackEventHighDefinitionUpdate];

            [[theValue(eventWasTriggered) should] beTrue];
        });

        it(@"should trigger its event after listen to playback's bit rate event with the respective params", ^{
            float expectedBitRate = 12.34;

            __block float bitRate = 0.0;
            [container once:CLPContainerEventBitRate callback:^(NSDictionary *userInfo) {
                bitRate = [userInfo[@"bit_rate"] floatValue];
            }];

            NSDictionary *userInfo = @{@"bit_rate": @(expectedBitRate)};
            [playback trigger:CLPPlaybackEventBitRate userInfo:userInfo];

            [[theValue(bitRate) should] equal:theValue(expectedBitRate)];
        });

        it(@"should trigger its event after listen to playback's state changed event", ^{
            __block BOOL eventWasTriggered = NO;
            [container once:CLPContainerEventPlaybackStateChanged callback:^(NSDictionary *userInfo) {
                eventWasTriggered = YES;
            }];

            [playback trigger:CLPPlaybackEventStateChanged];

            [[theValue(eventWasTriggered) should] beTrue];
        });

        it(@"should trigger its event after listen to playback's DVR state changed event with the respective params", ^{
            __block BOOL dvrInUse = NO;
            [container once:CLPContainerEventPlaybackStateDVRStateChanged callback:^(NSDictionary *userInfo) {
                dvrInUse = [userInfo[@"dvr_in_use"] boolValue];
            }];

            [playback trigger:CLPPlaybackEventDVRStateChanged userInfo:@{@"dvr_in_use": @(YES)}];

            [[theValue(dvrInUse) should] beTrue];
        });

        it(@"should update dvrInUse flag after listen to playback's DVR state changed event", ^{
            [[theValue(container.dvrInUse) should] beFalse];

            NSDictionary *userInfo = @{@"dvr_in_use": @(YES)};
            [playback trigger:CLPPlaybackEventDVRStateChanged userInfo:userInfo];

            [[theValue(container.dvrInUse) should] beTrue];
        });

        it(@"should update its settings after listen to playback's DVR state changed event", ^{
            [playback stub:@selector(settings) andReturn:@{@"foo": @"bar"}];

            [[container.settings should] beNil];

            [playback trigger:CLPPlaybackEventDVRStateChanged];

            [[container.settings should] equal:@{@"foo": @"bar"}];
        });

        it(@"should disable media control after listen to playback's disable media control event", ^{
            [playback trigger:CLPPlaybackEventMediaControlDisabled];
            [[theValue(container.mediaControlDisabled) should] beTrue];
        });

        it(@"should trigger its event after listen to playbacks's disable media control event", ^{
            __block BOOL eventWasTriggered = NO;
            [container once:CLPContainerEventMediaControlDisabled callback:^(NSDictionary *userInfo) {
                eventWasTriggered = YES;
            }];

            [playback trigger:CLPPlaybackEventMediaControlDisabled];

            [[theValue(eventWasTriggered) should] beTrue];
        });

        it(@"should enable media control after listen to playback's enable media control event", ^{
            // just to not get the default value and we think the test
            // result is what we're expecting
            [container setValue:@(YES) forKey:@"mediaControlDisabled"];

            [playback trigger:CLPPlaybackEventMediaControlEnabled];
            [[theValue(container.mediaControlDisabled) should] beFalse];
        });

        it(@"should trigger its event after listen to playbacks's enable media control event", ^{
            __block BOOL eventWasTriggered = NO;
            [container once:CLPContainerEventMediaControlEnabled callback:^(NSDictionary *userInfo) {
                eventWasTriggered = YES;
            }];

            [playback trigger:CLPPlaybackEventMediaControlEnabled];

            [[theValue(eventWasTriggered) should] beTrue];
        });

        it(@"should trigger its event after listen to playback's end event", ^{
            __block BOOL eventWasTriggered = NO;
            [container once:CLPContainerEventEnded callback:^(NSDictionary *userInfo) {
                eventWasTriggered = YES;
            }];

            [playback trigger:CLPPlaybackEventEnded];

            [[theValue(eventWasTriggered) should] beTrue];
        });

        it(@"should trigger its play event after listen to playback's play event", ^{
            __block BOOL eventWasTriggered = NO;
            [container once:CLPContainerEventPlay callback:^(NSDictionary *userInfo) {
                eventWasTriggered = YES;
            }];

            [playback trigger:CLPPlaybackEventPlay];

            [[theValue(eventWasTriggered) should] beTrue];
        });

        it(@"should trigger its pause event after listen to playback's pause event", ^{
            __block BOOL eventWasTriggered = NO;
            [container once:CLPContainerEventPause callback:^(NSDictionary *userInfo) {
                eventWasTriggered = YES;
            }];

            [playback trigger:CLPPlaybackEventPause];

            [[theValue(eventWasTriggered) should] beTrue];
        });

        it(@"should trigger its event after listen to playback's error event with respective params", ^{
            NSError *expectedErrorObject = [NSError new];

            __block NSError *error = nil;
            [container once:CLPContainerEventError callback:^(NSDictionary *userInfo) {
                error = expectedErrorObject;
            }];

            [playback trigger:CLPPlaybackEventError userInfo:@{@"error":expectedErrorObject}];

            [[error should] equal:expectedErrorObject];
        });
    });
});

SPEC_END

File: Example/Tests/ContainerTests.swift
import Quick
import Nimble
import Clappr

class ContainerTests: QuickSpec {
    
    override func spec() {
        describe("Container") {
            var container: Container!
            var playback: Playback!
            let sourceURL = NSURL(string: "http://globo.com/video.mp4")!
            
            beforeEach() {
                playback = Playback(url: sourceURL)
                container = Container(playback: playback)
            }
            
            describe("Initialization") {
                it("Should have the playback as subview") {
                    expect(playback.superview) == container
                }
            }
            
            describe("Destroy") {
                it("Should be removed from superview and destroy playback when destroy is called") {
                    let wrapperView = UIView()
                    wrapperView.addSubview(container)
                    
                    container.destroy()
                    
                    expect(playback.superview).to(beNil())
                    expect(container.superview).to(beNil())
                }
                
                it("Should stop listening to events after destroy is called") {
                    var callbackWasCalled = false
                    container.on("some-event") { _ in
                        callbackWasCalled = true
                    }
                    
                    container.destroy()
                    container.trigger("some-event")
                    
                    expect(callbackWasCalled) == false
                }
            }
            
            describe("Event Binding") {
                var eventWasTriggered = false
                let eventCallback: EventCallback = { _ in
                    eventWasTriggered = true
                }
                
                beforeEach{
                    eventWasTriggered = false
                }
                
                
                it("Should trigger container progress event when playback progress event happens") {
                    let expectedStart: Float = 0.7, expectedEnd: Float = 15.4, expectedDuration: NSTimeInterval = 10
                    var start: Float!, end: Float!, duration: NSTimeInterval!
                    
                    container.once(ContainerEvent.Progress.rawValue) { userInfo in
                        start = userInfo?["start_position"] as! Float
                        end = userInfo?["end_position"] as! Float
                        duration = userInfo?["duration"] as! NSTimeInterval
                    }
                    
                    let userInfo: EventUserInfo = ["start_position": expectedStart,
                        "end_position": expectedEnd,
                        "duration": expectedDuration]
                    playback.trigger(PlaybackEvent.Progress.rawValue, userInfo: userInfo)
                    
                    expect(start) == expectedStart
                    expect(end) == expectedEnd
                    expect(duration) == expectedDuration
                }
                
                it("Should trigger container time updated event when playback respective event happens") {
                    let expectedPosition: Float = 10.3, expectedDuration: NSTimeInterval = 12.7
                    var position: Float!, duration: NSTimeInterval!
                    
                    container.once(ContainerEvent.TimeUpdated.rawValue) { userInfo in
                        position = userInfo?["position"] as! Float
                        duration = userInfo?["duration"] as! NSTimeInterval
                    }
                    
                    let userInfo: EventUserInfo = ["position": expectedPosition, "duration": expectedDuration]
                    playback.trigger(PlaybackEvent.TimeUpdated.rawValue, userInfo: userInfo)
                    
                    expect(position) == expectedPosition
                    expect(duration) == expectedDuration
                }
                
                it("Should trigger container loaded metadata event when playback respective event happens") {
                    let expectedDuration: NSTimeInterval = 20.0
                    var duration: NSTimeInterval!
                    
                    container.once(ContainerEvent.LoadedMetadata.rawValue) { userInfo in
                        duration = userInfo?["duration"] as! NSTimeInterval
                    }
                    
                    let userInfo: EventUserInfo = ["duration": expectedDuration]
                    playback.trigger(PlaybackEvent.LoadedMetadata.rawValue, userInfo: userInfo)
                    
                    expect(duration) == expectedDuration
                }
                
                it("Should trigger container bit rate event when playback respective event happens") {
                    let expectedBitRate: NSTimeInterval = 11.0
                    var bitRate: NSTimeInterval!
                    
                    container.once(ContainerEvent.BitRate.rawValue) { userInfo in
                        bitRate = userInfo?["bit_rate"] as! NSTimeInterval
                    }
                    
                    let userInfo: EventUserInfo = ["bit_rate": expectedBitRate]
                    playback.trigger(PlaybackEvent.BitRate.rawValue, userInfo: userInfo)
                    
                    expect(bitRate) == expectedBitRate
                }
                
                it("Should trigger container DVR state event when playback respective event happens with params") {
                    var dvrInUse = false
                    
                    container.once(ContainerEvent.PlaybackDVRStateChanged.rawValue) { userInfo in
                        dvrInUse = userInfo?["dvr_in_use"] as! Bool
                    }
                    
                    let userInfo: EventUserInfo = ["dvr_in_use": true]
                    playback.trigger(PlaybackEvent.DVRStateChanged.rawValue, userInfo: userInfo)
                    
                    expect(dvrInUse).to(beTrue())
                }
                
                it("Should trigger container Error event when playback respective event happens with params") {
                    var error = ""
                    
                    container.once(ContainerEvent.Error.rawValue) { userInfo in
                        error = userInfo?["error"] as! String
                    }
                    
                    let userInfo: EventUserInfo = ["error": "Error"]
                    playback.trigger(PlaybackEvent.Error.rawValue, userInfo: userInfo)
                    
                    expect(error) == "Error"
                }
                
                it("Should update container dvrInUse property on playback DVRSTateChanged event") {
                    let userInfo: EventUserInfo = ["dvr_in_use": true]
                    
                    expect(container.dvrInUse).to(beFalse())
                    playback.trigger(PlaybackEvent.DVRStateChanged.rawValue, userInfo: userInfo)
                    expect(container.dvrInUse).to(beTrue())
                }
                
                it("Should be ready after playback ready event is triggered") {
                    expect(container.ready) == false
                    playback.trigger(PlaybackEvent.Ready.rawValue)
                    expect(container.ready) == true
                }
                
                it("Should trigger buffering event after playback respective event is triggered") {
                    container.on(ContainerEvent.Buffering.rawValue, callback: eventCallback)
                    playback.trigger(PlaybackEvent.Buffering.rawValue)
                    expect(eventWasTriggered) == true
                }
                
                it("Should trigger buffer full event after playback respective event is triggered") {
                    container.on(ContainerEvent.BufferFull.rawValue, callback: eventCallback)
                    playback.trigger(PlaybackEvent.BufferFull.rawValue)
                    expect(eventWasTriggered) == true
                }
                
                it("Should trigger settings event after playback respective event is triggered") {
                    container.on(ContainerEvent.SettingsUpdated.rawValue, callback: eventCallback)
                    playback.trigger(PlaybackEvent.SettingsUpdated.rawValue)
                    expect(eventWasTriggered) == true
                }
                
                it("Should trigger HD updated event after playback respective event is triggered") {
                    container.on(ContainerEvent.HighDefinitionUpdated.rawValue, callback: eventCallback)
                    playback.trigger(PlaybackEvent.HighDefinitionUpdated.rawValue)
                    expect(eventWasTriggered) == true
                }
                
                it("Should trigger State Changed event after playback respective event is triggered") {
                    container.on(ContainerEvent.PlaybackStateChanged.rawValue, callback: eventCallback)
                    playback.trigger(PlaybackEvent.StateChanged.rawValue)
                    expect(eventWasTriggered) == true
                }
                
                it("Should trigger Media Control Disabled event after playback respective event is triggered") {
                    container.on(ContainerEvent.MediaControlDisabled.rawValue, callback: eventCallback)
                    playback.trigger(PlaybackEvent.MediaControlDisabled.rawValue)
                    expect(eventWasTriggered) == true
                }

                it("Should trigger Media Control Enabled event after playback respective event is triggered") {
                    container.on(ContainerEvent.MediaControlEnabled.rawValue, callback: eventCallback)
                    playback.trigger(PlaybackEvent.MediaControlEnabled.rawValue)
                    expect(eventWasTriggered) == true
                }
                
                it("Should update mediaControlEnabled property after playback MediaControleEnabled or Disabled is triggered") {
                    playback.trigger(PlaybackEvent.MediaControlEnabled.rawValue)
                    expect(container.mediaControlEnabled).to(beTrue())
                    playback.trigger(PlaybackEvent.MediaControlDisabled.rawValue)
                    expect(container.mediaControlEnabled).to(beFalse())
                }
                
                it("Should trigger Ended event after playback respective event is triggered") {
                    container.on(ContainerEvent.Ended.rawValue, callback: eventCallback)
                    playback.trigger(PlaybackEvent.Ended.rawValue)
                    expect(eventWasTriggered) == true
                }
                
                it("Should trigger Play event after playback respective event is triggered") {
                    container.on(ContainerEvent.Play.rawValue, callback: eventCallback)
                    playback.trigger(PlaybackEvent.Play.rawValue)
                    expect(eventWasTriggered) == true
                }
                
                it("Should trigger Pause event after playback respective event is triggered") {
                    container.on(ContainerEvent.Pause.rawValue, callback: eventCallback)
                    playback.trigger(PlaybackEvent.Pause.rawValue)
                    expect(eventWasTriggered) == true
                }
                
                it("Should trigger it's Stop event after stop is called") {
                    container.on(ContainerEvent.Stop.rawValue, callback: eventCallback)
                    container.stop()
                    expect(eventWasTriggered) == true
                }
                
                context("Bindings with mocked playback") {
                    class MockedSettingsPlayback: Playback {
                        var stopWasCalled = false , playWasCalled = false, pauseWasCalled = false
                        override var settings: [String: AnyObject] {
                            get {
                                return ["foo": "bar"]
                            }
                        }
                        
                        override private func stop() {
                            stopWasCalled = true
                        }
                        
                        override private func pause() {
                            pauseWasCalled = true
                        }
                        
                        private override func play() {
                            playWasCalled = true
                        }
                    }
                    
                    var mockedPlayback: MockedSettingsPlayback!
                    
                    beforeEach() {
                        mockedPlayback = MockedSettingsPlayback(url: sourceURL)
                        container = Container(playback: mockedPlayback)
                    }
                    
                    it("Should update it's settings after playback's settings update event") {
                        mockedPlayback.trigger(PlaybackEvent.SettingsUpdated.rawValue)
                        let fooSetting = container.settings["foo"] as? String
                        expect(fooSetting) == "bar"
                    }
                    
                    it("Should update it's settings after playback's DVR State changed event") {
                        mockedPlayback.trigger(PlaybackEvent.DVRStateChanged.rawValue)
                        let fooSetting = container.settings["foo"] as? String
                        expect(fooSetting) == "bar"
                    }
                    
                    it("Should call playback's stop method after calling respective method on container") {
                        container.stop()
                        expect(mockedPlayback.stopWasCalled).to(beTrue())
                    }
                    
                    it("Should call playback's play method after calling respective method on container") {
                        container.play()
                        expect(mockedPlayback.playWasCalled).to(beTrue())
                    }
                    
                    it("Should call playback's pause method after calling respective method on container") {
                        container.pause()
                        expect(mockedPlayback.pauseWasCalled).to(beTrue())
                    }
                }
                
            }
        }
    }
}